sqlccl: don't split CSV files mid column family

madelynnblue opened the pull request:
Previously, we would allow splitting between two KVs in the index of
different column families. This caused numerous odd errors, including
double-counting these rows during RESTORE. The select inconsistencies
mentioned in the fixed issues are also explained by this, I think,
because the system isn't quite sure what to do with these partial
rows and so does some weird stuff.

Fixes #18076

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18433)
<!-- Reviewable:end -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 518 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-KuL6HHvLa4bwPDsZI24:-KuL6HHvLa4bwPDsZI25:betepba) ([raw file](https://github.com/cockroachdb/cockroach/blob/964a2cac4e61ebe12401ba4f31607295794b7040/pkg/ccl/sqlccl/csv.go#L518)):*
> ```Go
> 	// current SST hasn't yet filled up. Once the SST has filled up, lastKey is
> 	// set to the key at which to stop adding KVs. We have to do this because
> 	// column families cannot be split among SSTs and the SST may have filled
> ```

nit: "all column families for a row must be in one SST"

---

*[pkg/ccl/sqlccl/csv.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-KuL6mD_NPRRncwbnYjU:-KuL6mD_NPRRncwbnYjV:boxrtne) ([raw file](https://github.com/cockroachdb/cockroach/blob/964a2cac4e61ebe12401ba4f31607295794b7040/pkg/ccl/sqlccl/csv.go#L559)):*
> ```Go
> 			return 0, errors.Wrapf(err, errSSTCreationMaybeDuplicateTemplate, kv.Key.Key)
> 		}
> 		if sst.DataSize > sstMaxSize {
> ```

could combine the two ifs: `if sst.DataSize > sstMaxSize && lastKey == nil {` 

---

*[pkg/ccl/sqlccl/csv_test.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-KuL7-RuHLvO1W_H7SiY:-KuL7-RuHLvO1W_H7SiZ:b-3rbfl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/964a2cac4e61ebe12401ba4f31607295794b7040/pkg/ccl/sqlccl/csv_test.go#L348)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	csv, kv, sst, err := sqlccl.LoadCSV(ctx, tablePath, []string{dataPath}, tmp, 0 /* comma */, 0 /* comment */, nil /* nullif */, 1 /* sstMaxSize */, tmp)
> ```

comment explaining that sstMaxSize = 1 would naturally put each kv in it's own sst

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-:-KuL76bN4o0ChuAD96eQ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sqlccl: don't split CSV files mid column family

Previously, we would allow splitting between two KVs in the index of
different column families. This caused numerous odd errors, including
double-counting these rows during RESTORE. The select inconsistencies
mentioned in the fixed issues are also explained by this, I think,
because the system isn't quite sure what to do with these partial
rows and so does some weird stuff.

Fixes #18076

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 518 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-KuL6HHvLa4bwPDsZI24:-KuLCIRGV-vvMn1HNUq0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/964a2cac4e61ebe12401ba4f31607295794b7040/pkg/ccl/sqlccl/csv.go#L518)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "all column families for a row must be in one SST"
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv.go, line 559 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-KuL6mD_NPRRncwbnYjU:-KuLCJ36yUB8ApDIZU-4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/964a2cac4e61ebe12401ba4f31607295794b7040/pkg/ccl/sqlccl/csv.go#L559)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

could combine the two ifs: `if sst.DataSize > sstMaxSize && lastKey == nil {` 
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/csv_test.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433#-KuL7-RuHLvO1W_H7SiY:-KuLCevSIveWnt3GosYM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/964a2cac4e61ebe12401ba4f31607295794b7040/pkg/ccl/sqlccl/csv_test.go#L348)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

comment explaining that sstMaxSize = 1 would naturally put each kv in it's own sst
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18433)*
<!-- Sent from Reviewable.io -->
